File: tests/functional/Bots/DotNet/Consumers/CodeFirst/SimpleHostBot-2.1/Bots/HostBot.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Builder.Integration.AspNet.Core.Skills;
using Microsoft.Bot.Builder.Skills;
using Microsoft.Bot.Connector.Authentication;
using Microsoft.Bot.Schema;
using Microsoft.BotFrameworkFunctionalTests.SimpleHostBot21.Dialogs;
using Microsoft.Extensions.Configuration;
using Newtonsoft.Json;

namespace Microsoft.BotFrameworkFunctionalTests.SimpleHostBot21.Bots
{
    public class HostBot : ActivityHandler
    {
        public const string DeliveryModePropertyName = "deliveryModeProperty";
        public const string ActiveSkillPropertyName = "activeSkillProperty";

        private readonly IStatePropertyAccessor<string> _deliveryModeProperty;
        private readonly IStatePropertyAccessor<BotFrameworkSkill> _activeSkillProperty;
        private readonly IStatePropertyAccessor<DialogState> _dialogStateProperty;
        private readonly string _botId;
        private readonly ConversationState _conversationState;
        private readonly SkillHttpClient _skillClient;
        private readonly SkillsConfiguration _skillsConfig;
        private readonly Dialog _dialog;

        /// <summary>
        /// Initializes a new instance of the <see cref="HostBot"/> class.
        /// </summary>
        /// <param name="conversationState">A state management object for the conversation.</param>
        /// <param name="skillsConfig">The skills configuration.</param>
        /// <param name="skillClient">The HTTP client for the skills.</param>
        /// <param name="configuration">The configuration properties.</param>
        /// <param name="dialog">The dialog to use.</param>
        public HostBot(ConversationState conversationState, SkillsConfiguration skillsConfig, SkillHttpClient skillClient, IConfiguration configuration, SetupDialog dialog)
        {
            _conversationState = conversationState ?? throw new ArgumentNullException(nameof(conversationState));
            _skillsConfig = skillsConfig ?? throw new ArgumentNullException(nameof(skillsConfig));
            _skillClient = skillClient ?? throw new ArgumentNullException(nameof(skillClient));
            _dialog = dialog ?? throw new ArgumentNullException(nameof(dialog));
            _dialogStateProperty = _conversationState.CreateProperty<DialogState>("DialogState");
            if (configuration == null)
            {
                throw new ArgumentNullException(nameof(configuration));
            }

            _botId = configuration.GetSection(MicrosoftAppCredentials.MicrosoftAppIdKey)?.Value;

            // Create state properties to track the delivery mode and active skill.
            _deliveryModeProperty = conversationState.CreateProperty<string>(DeliveryModePropertyName);
            _activeSkillProperty = conversationState.CreateProperty<BotFrameworkSkill>(ActiveSkillPropertyName);
        }

        /// <inheritdoc/>
        public override async Task OnTurnAsync(ITurnContext turnContext, CancellationToken cancellationToken = default)
        {
            // Forward all activities except EndOfConversation to the active skill.
            if (turnContext.Activity.Type != ActivityTypes.EndOfConversation)
            {
                // Try to get the active skill
                var activeSkill = await _activeSkillProperty.GetAsync(turnContext, () => null, cancellationToken);

                if (activeSkill != null)
                {
                    var deliveryMode = await _deliveryModeProperty.GetAsync(turnContext, () => null, cancellationToken);

                    // Send the activity to the skill
                    await SendToSkillAsync(turnContext, deliveryMode, activeSkill, cancellationToken);
                    return;
                }
            }

            await base.OnTurnAsync(turnContext, cancellationToken);

            // Save any state changes that might have occurred during the turn.
            await _conversationState.SaveChangesAsync(turnContext, false, cancellationToken);
        }

        /// <summary>
        /// Processes a message activity.
        /// </summary>
        /// <param name="turnContext">Context for the current turn of conversation.</param>
        /// <param name="cancellationToken">CancellationToken propagates notifications that operations should be cancelled.</param>
        /// <returns>A <see cref="Task{TResult}"/> representing the result of the asynchronous operation.</returns>
        protected override async Task OnMessageActivityAsync(ITurnContext<IMessageActivity> turnContext, CancellationToken cancellationToken)
        {
            if (_skillsConfig.Skills.ContainsKey(turnContext.Activity.Text))
            {
                var deliveryMode = await _deliveryModeProperty.GetAsync(turnContext, () => null, cancellationToken);
                var selectedSkill = _skillsConfig.Skills[turnContext.Activity.Text];
                var v3Bots = new List<string> { "EchoSkillBotDotNetV3", "EchoSkillBotJSV3" };

                if (selectedSkill != null && deliveryMode == DeliveryModes.ExpectReplies && v3Bots.Contains(selectedSkill.Id))
                {
                    var message = MessageFactory.Text("V3 Bots do not support 'expectReplies' delivery mode.");
                    await turnContext.SendActivityAsync(message, cancellationToken);

                    // Forget delivery mode and skill invocation.
                    await _deliveryModeProperty.DeleteAsync(turnContext, cancellationToken);
                    await _activeSkillProperty.DeleteAsync(turnContext, cancellationToken);

                    // Restart setup dialog
                    await _conversationState.DeleteAsync(turnContext, cancellationToken);
                }
            }

            await _dialog.RunAsync(turnContext, _dialogStateProperty, cancellationToken);
        }

        /// <summary>
        /// Processes an end of conversation activity.
        /// </summary>
        /// <param name="turnContext">Context for the current turn of conversation.</param>
        /// <param name="cancellationToken">CancellationToken propagates notifications that operations should be cancelled.</param>
        /// <returns>A <see cref="Task{TResult}"/> representing the result of the asynchronous operation.</returns>
        protected override async Task OnEndOfConversationActivityAsync(ITurnContext<IEndOfConversationActivity> turnContext, CancellationToken cancellationToken)
        {
            await EndConversation((Activity)turnContext.Activity, turnContext, cancellationToken);
        }

        /// <summary>
        /// Processes a member added event.
        /// </summary>
        /// <param name="membersAdded">The list of members added to the conversation.</param>
        /// <param name="turnContext">Context for the current turn of conversation.</param>
        /// <param name="cancellationToken">CancellationToken propagates notifications that operations should be cancelled.</param>
        /// <returns>A <see cref="Task{TResult}"/> representing the result of the asynchronous operation.</returns>
        protected override async Task OnMembersAddedAsync(IList<ChannelAccount> membersAdded, ITurnContext<IConversationUpdateActivity> turnContext, CancellationToken cancellationToken)
        {
            foreach (var member in membersAdded)
            {
                if (member.Id != turnContext.Activity.Recipient.Id)
                {
                    await turnContext.SendActivityAsync(MessageFactory.Text("Hello and welcome!"), cancellationToken);
                    await _dialog.RunAsync(turnContext, _dialogStateProperty, cancellationToken);
                }
            }
        }

        /// <summary>
        /// Clears storage variables and sends the end of conversation activities.
        /// </summary>
        /// <param name="activity">End of conversation activity.</param>
        /// <param name="turnContext">Context for the current turn of conversation.</param>
        /// <param name="cancellationToken">CancellationToken propagates notifications that operations should be cancelled.</param>
        /// <returns>A <see cref="Task{TResult}"/> representing the result of the asynchronous operation.</returns>
        private async Task EndConversation(Activity activity, ITurnContext turnContext, CancellationToken cancellationToken)
        {
            // Forget delivery mode and skill invocation.
            await _deliveryModeProperty.DeleteAsync(turnContext, cancellationToken);
            await _activeSkillProperty.DeleteAsync(turnContext, cancellationToken);

            // Show status message, text and value returned by the skill
            var eocActivityMessage = $"Received {ActivityTypes.EndOfConversation}.\n\nCode: {activity.Code}.";
            if (!string.IsNullOrWhiteSpace(activity.Text))
            {
                eocActivityMessage += $"\n\nText: {activity.Text}";
            }

            if (activity.Value != null)
            {
                eocActivityMessage += $"\n\nValue: {JsonConvert.SerializeObject(activity.Value)}";
            }

            await turnContext.SendActivityAsync(MessageFactory.Text(eocActivityMessage), cancellationToken);

            // We are back at the host.
            await turnContext.SendActivityAsync(MessageFactory.Text("Back in the host bot."), cancellationToken);

            // Restart setup dialog.
            await _dialog.RunAsync(turnContext, _dialogStateProperty, cancellationToken);

            await _conversationState.SaveChangesAsync(turnContext, false, cancellationToken);
        }

        /// <summary>
        /// Sends an activity to the skill bot.
        /// </summary>
        /// <param name="turnContext">Context for the current turn of conversation.</param>
        /// <param name="deliveryMode">The delivery mode to use when communicating to the skill.</param>
        /// <param name="targetSkill">The skill that will receive the activity.</param>
        /// <param name="cancellationToken">CancellationToken propagates notifications that operations should be cancelled.</param>
        /// <returns>A <see cref="Task{TResult}"/> representing the result of the asynchronous operation.</returns>
        private async Task SendToSkillAsync(ITurnContext turnContext, string deliveryMode, BotFrameworkSkill targetSkill, CancellationToken cancellationToken)
        {
            // NOTE: Always SaveChanges() before calling a skill so that any activity generated by the skill
            // will have access to current accurate state.
            await _conversationState.SaveChangesAsync(turnContext, force: true, cancellationToken: cancellationToken);

            if (deliveryMode == DeliveryModes.ExpectReplies)
            {
                // Clone activity and update its delivery mode.
                var activity = JsonConvert.DeserializeObject<Activity>(JsonConvert.SerializeObject(turnContext.Activity));
                activity.DeliveryMode = deliveryMode;

                // Route the activity to the skill.
                var expectRepliesResponse = await _skillClient.PostActivityAsync<ExpectedReplies>(_botId, targetSkill, _skillsConfig.SkillHostEndpoint, activity, cancellationToken);

                // Check response status.
                if (!expectRepliesResponse.IsSuccessStatusCode())
                {
                    throw new HttpRequestException($"Error invoking the skill id: \"{targetSkill.Id}\" at \"{targetSkill.SkillEndpoint}\" (status is {expectRepliesResponse.Status}). \r\n {expectRepliesResponse.Body}");
                }

                // Route response activities back to the channel.
                var responseActivities = expectRepliesResponse.Body?.Activities;

                foreach (var responseActivity in responseActivities)
                {
                    if (responseActivity.Type == ActivityTypes.EndOfConversation)
                    {
                        await EndConversation(responseActivity, turnContext, cancellationToken);
                    }
                    else
                    {
                        await turnContext.SendActivityAsync(responseActivity, cancellationToken);
                    }
                }
            }
            else
            {
                // Route the activity to the skill.
                var response = await _skillClient.PostActivityAsync(_botId, targetSkill, _skillsConfig.SkillHostEndpoint, (Activity)turnContext.Activity, cancellationToken);

                // Check response status
                if (!response.IsSuccessStatusCode())
                {
                    throw new HttpRequestException($"Error invoking the skill id: \"{targetSkill.Id}\" at \"{targetSkill.SkillEndpoint}\" (status is {response.Status}). \r\n {response.Body}");
                }
            }
        }
    }
}

File: tests/functional/Bots/Python/Consumers/CodeFirst/SimpleHostBot/bots/host_bot.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.
import copy
from typing import List
from botbuilder.core import (
    ActivityHandler,
    ConversationState,
    MessageFactory,
    TurnContext,
)
from botbuilder.core.skills import BotFrameworkSkill
from botbuilder.dialogs import Dialog
from botbuilder.schema import (
    ActivityTypes,
    ChannelAccount,
    ExpectedReplies,
    DeliveryModes,
)
from botbuilder.integration.aiohttp.skills import SkillHttpClient

from config import DefaultConfig, SkillConfiguration
from helpers.dialog_helper import DialogHelper

DELIVERY_MODE_PROPERTY_NAME = "deliveryModeProperty"
ACTIVE_SKILL_PROPERTY_NAME = "activeSkillProperty"


class HostBot(ActivityHandler):
    def __init__(
        self,
        conversation_state: ConversationState,
        skills_config: SkillConfiguration,
        skill_client: SkillHttpClient,
        config: DefaultConfig,
        dialog: Dialog,
    ):
        self._bot_id = config.APP_ID
        self._skill_client = skill_client
        self._skills_config = skills_config
        self._conversation_state = conversation_state
        self._dialog = dialog
        self._dialog_state_property = conversation_state.create_property("DialogState")

        # Create state property to track the delivery mode and active skill.
        self._delivery_mode_property = conversation_state.create_property(
            DELIVERY_MODE_PROPERTY_NAME
        )
        self._active_skill_property = conversation_state.create_property(
            ACTIVE_SKILL_PROPERTY_NAME
        )

    async def on_turn(self, turn_context):
        # Forward all activities except EndOfConversation to the active skill.
        if turn_context.activity.type != ActivityTypes.end_of_conversation:
            # Try to get the active skill
            active_skill: BotFrameworkSkill = await self._active_skill_property.get(
                turn_context
            )

            if active_skill:
                delivery_mode: str = await self._delivery_mode_property.get(
                    turn_context
                )

                # Send the activity to the skill
                await self.__send_to_skill(turn_context, delivery_mode, active_skill)
                return

        await super().on_turn(turn_context)
        # Save any state changes that might have occurred during the turn.
        await self._conversation_state.save_changes(turn_context)

    async def on_message_activity(self, turn_context: TurnContext):
        if turn_context.activity.text in self._skills_config.SKILLS:
            delivery_mode: str = await self._delivery_mode_property.get(turn_context)
            selected_skill = self._skills_config.SKILLS[turn_context.activity.text]
            v3_bots = ["EchoSkillBotDotNetV3", "EchoSkillBotJSV3"]

            if (
                selected_skill
                and delivery_mode == DeliveryModes.expect_replies
                and selected_skill.id.lower() in (id.lower() for id in v3_bots)
            ):
                message = MessageFactory.text(
                    "V3 Bots do not support 'expectReplies' delivery mode."
                )
                await turn_context.send_activity(message)

                # Forget delivery mode and skill invocation.
                await self._delivery_mode_property.delete(turn_context)

                # Restart setup dialog
                await self._conversation_state.delete(turn_context)

        await DialogHelper.run_dialog(
            self._dialog,
            turn_context,
            self._dialog_state_property,
        )

    async def on_end_of_conversation_activity(self, turn_context: TurnContext):
        await self.end_conversation(turn_context.activity, turn_context)

    async def on_members_added_activity(
        self, members_added: List[ChannelAccount], turn_context: TurnContext
    ):
        for member in members_added:
            if member.id != turn_context.activity.recipient.id:
                await turn_context.send_activity(
                    MessageFactory.text("Hello and welcome!")
                )
                await DialogHelper.run_dialog(
                    self._dialog,
                    turn_context,
                    self._dialog_state_property,
                )

    async def end_conversation(self, activity, turn_context):
        # Forget delivery mode and skill invocation.
        await self._delivery_mode_property.delete(turn_context)
        await self._active_skill_property.delete(turn_context)

        eoc_activity_message = (
            f"Received {ActivityTypes.end_of_conversation}.\n\nCode: {activity.code}."
        )
        if activity.text:
            eoc_activity_message = eoc_activity_message + f"\n\nText: {activity.text}"
        if activity.value:
            eoc_activity_message = eoc_activity_message + f"\n\nValue: {activity.value}"
        await turn_context.send_activity(eoc_activity_message)

        # We are back
        await turn_context.send_activity(MessageFactory.text("Back in the host bot."))

        # Restart setup dialog.
        await DialogHelper.run_dialog(
            self._dialog,
            turn_context,
            self._dialog_state_property,
        )

        await self._conversation_state.save_changes(turn_context)

    async def __send_to_skill(
        self,
        turn_context: TurnContext,
        delivery_mode: str,
        target_skill: BotFrameworkSkill,
    ):
        # NOTE: Always SaveChanges() before calling a skill so that any activity generated by the skill
        # will have access to current accurate state.
        await self._conversation_state.save_changes(turn_context, force=True)

        if delivery_mode == "expectReplies":
            # Clone activity and update its delivery mode.
            activity = copy.copy(turn_context.activity)
            activity.delivery_mode = delivery_mode

            # Route the activity to the skill.
            expect_replies_response = await self._skill_client.post_activity_to_skill(
                self._bot_id,
                target_skill,
                self._skills_config.SKILL_HOST_ENDPOINT,
                activity,
            )

            # Route response activities back to the channel.
            response_activities: ExpectedReplies = (
                ExpectedReplies().deserialize(expect_replies_response.body).activities
            )

            for response_activity in response_activities:
                if response_activity.type == ActivityTypes.end_of_conversation:
                    await self.end_conversation(response_activity, turn_context)

                else:
                    await turn_context.send_activity(response_activity)

        else:
            # Route the activity to the skill.
            await self._skill_client.post_activity_to_skill(
                self._bot_id,
                target_skill,
                self._skills_config.SKILL_HOST_ENDPOINT,
                turn_context.activity,
            )
